README.md
# Netlify Plugin Encrypted Files

[Netlify Build](https://github.com/netlify/build) Plugin to partially obscure files (names and contents) in git repos!

This enables you to ***partially* open source your site**, while still being able to work as normal on your local machine and in your Netlify builds. 

External viewers of your Git repo will just see hashes for your secret content while it is encrypted - and as long as it still builds without these files, they can still contribute to the rest of your site.

## Demo

https://netlify-plugin-encrypted-files.netlify.com

## Usage

To install, add the following lines to your `netlify.toml` file:

```toml
[[plugins]]
package = "netlify-plugin-encrypted-files"
  # all inputs are optional. uncomment to apply
  # [plugins.inputs]
  # branches = [ # if specified, allow a small set of branches for which the decrypt is applied
    # "master",
    # "swyx/myNewBranch"
  # ] # dont forget to specify a NETLIFY_ENCRYPT_KEY env variable in Netlify's UI
```

In your local environment, install the plugin and run the `encrypt` CLI on your project specifying a glob filepath for what should be encrypted and what `NETLIFY_ENCRYPT_KEY` you intend to use, e.g.

```bash
npm i netlify-plugin-encrypted-files
NETLIFY_ENCRYPT_KEY='test' node encrypt.js content/secretstuff/**/*.*
```

This generates a `.encrypted` folder which you should check into git. ([Here's how it looks - even filenames are obscured!](https://github.com/sw-yx/netlify-plugin-encrypted-tree/master/.encrypted))

Also dont forget to `.gitignore` your secret content!

On Netlify's side, all it does is it runs `decrypt` for you, using the same `NETLIFY_ENCRYPT_KEY` you used to encrypt it. To set the environment variable without it being visible in git, you should [use the Netlify UI](https://docs.netlify.com/configure-builds/environment-variables/#declare-variables).

![image](https://user-images.githubusercontent.com/6764957/77190041-e827d600-6aae-11ea-9a5d-5dc13b4b0475.png)


## How It Works

This plugin is an unusual one: it has a CLI that works outside of the build bot, that you use to encrypt stuff. Then, inside the plugin, it runs the decrypt command for you before proceeding with the build.

The idea is:

1. while developing, work with your files as normal

- before committing, run `encrypt secretcontent/**/*.md` (any file matching logic here will do)
- make sure files-to-be-encrypted are gitignored
- `encrypt` will encrypt your files to the `.encrypt` folder with the `NETLIFY_ENCRYPT_KEY` environment variable
- so you run something like:
  -  `NETLIFY_ENCRYPT_KEY='test' yarn encrypt demo/secretstuff/**/*.*`
  -  or `NETLIFY_ENCRYPT_KEY='test' /node_modules/.bin/encrypt demo/secretstuff/**/*.*`
- check the new `.encrypt` folder into git

2. while deploying, this plugin runs a `decrypt` before any build and decrypts it with the same env variable
3. for collaborators, they should run `decrypt` on git pull. 
- so you run something like:
  -  `NETLIFY_ENCRYPT_KEY='test' yarn decrypt demo/secretstuff/**/*.*`
  -  or `NETLIFY_ENCRYPT_KEY='test' /node_modules/.bin/decrypt demo/secretstuff/**/*.*`
- NOTE: By default this overwrites files since that is usually the desired behavior, but if you want to be extra sure, you can add a `--testdecrypt` flag:
  -  `NETLIFY_ENCRYPT_KEY='test' yarn decrypt --testdecrypt demo/secretstuff/**/*.*`
  -  or `NETLIFY_ENCRYPT_KEY='test' /node_modules/.bin/decrypt --testdecrypt demo/secretstuff/**/*.*`
  - this will decrypt to a `testdecrypt` folder instead of the real destination, so you can preview what the effect of decrypting will be.

## Configuration

No configuration is required - by default the `decrypt`ing works on all Netlify Builds, but you can restrict it to a small set of branches you specify:

```toml
# netlify.toml
[[plugins]]
package = "netlify-plugin-encrypted-files"

  [plugins.inputs]
  
  # if specified, allow a small set of branches for which the decrypt is applied
  branches = [
    "master",
    "swyx/myNewBranch"
  ] 
  
  # dont forget to specify a NETLIFY_ENCRYPT_KEY env variable in Netlify's UI
```

## For Collaborators

To test this repo locally you can run:

-  `NETLIFY_ENCRYPT_KEY='test' node encrypt demo/secretstuff/**/*.*`
-  `NETLIFY_ENCRYPT_KEY='test' node decrypt --testdecrypt` by default decrypt


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v0.0.5](https://github.com/sw-yx/netlify-plugin-encrypted-files/compare/v0.0.4...v0.0.5) - 2020-03-20

### Commits

- demo [`34189f6`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/34189f61c0ccaafaebed0964beb71c65c9aba174)
- plugindecrypt [`9f70182`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/9f701823715d8493247962578010794a0ec2ffa3)
- changes [`d82770f`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/d82770fdda3f72d353949870b6456398ec0228d2)

## [v0.0.4](https://github.com/sw-yx/netlify-plugin-encrypted-files/compare/v0.0.3...v0.0.4) - 2020-03-18

### Merged

- Update installation instructions [`#2`](https://github.com/sw-yx/netlify-plugin-encrypted-files/pull/2)

### Commits

- Update README.md [`a0765e6`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/a0765e688ec7dc0aa95b88291c06554fddd27dd0)
- Update README.md [`db095ab`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/db095abe5be41b3eaf27a3618fc03d545a91c15b)
- Update README.md [`698a90e`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/698a90ea7876da555d745a3f0421b1e082133a62)

## [v0.0.3](https://github.com/sw-yx/netlify-plugin-encrypted-files/compare/v0.0.2...v0.0.3) - 2020-03-11

### Commits

- add branches feature [`0c1f9be`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/0c1f9bec49adfded3a43535b3c57d1ea151f3525)

## v0.0.2 - 2020-03-11

### Merged

- Update event handler name [`#1`](https://github.com/sw-yx/netlify-plugin-encrypted-files/pull/1)

### Commits

- add tests [`8142f7b`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/8142f7b072335194c3ea812a6a0d5072ddcc8a2c)
- update api's [`f2d566f`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/f2d566f5d8e904cc4c011b12573062217f9bf595)
- init [`a0dcc1e`](https://github.com/sw-yx/netlify-plugin-encrypted-files/commit/a0dcc1e954a958046846fd695145220bc36fa32b)


fixtures/encrypt.test.js
const fs = require('fs');
const path = require('path');

// actual test
const netlifyPlugin = require('../index.js');
test('plugin fixture works', async () => {
  expect(async () => {
    process.env.NETLIFY_ENCRYPT_KEY = 'test';
    const initPlugin = netlifyPlugin();
    console.log(`running ${initPlugin.name}`);
    initPlugin.onInit();
  }).not.toThrow();
});


pluginDecrypt.js
const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const zlib = require('zlib');

const ENCRYPTED_PLUGIN_SALT = 'ENCRYPTED_PLUGIN_SALT';
const { NETLIFY_ENCRYPT_KEY } = process.env;

module.exports = function pluginDecrypt({
  // unzip to '.testdecrypt' folder instead of overwriting real files
  testdecrypt = false
}) {
  if (typeof NETLIFY_ENCRYPT_KEY === 'undefined') {
    console.error(
      'must define NETLIFY_ENCRYPT_KEY to use netlify-plugin-encrypted-files. For more info: https://github.com/sw-yx/netlify-plugin-encrypted-files#usage'
    );
    process.exit(1);
  }
  const files = fs.readdirSync('.encrypted');
  files.forEach((sourceFilePath) => {
    let destinationfilePath = Buffer.from(sourceFilePath, 'base64').toString();
    decrypt(
      path.join('.encrypted', sourceFilePath),
      destinationfilePath,
      testdecrypt
    );
  });
};

// core crypto stuff
// https://medium.com/@brandonstilson/lets-encrypt-files-with-node-85037bea8c0e

function decrypt(sourceFilePath, destinationfilePath, testdecrypt) {
  // First, get the initialization vector from the file.
  const readInitVect = fs.createReadStream(sourceFilePath, { end: 15 });

  let initVect;
  readInitVect.on('data', (chunk) => {
    initVect = chunk;
  });

  // Once we’ve got the initialization vector, we can decrypt the file.
  readInitVect.on('close', () => {
    const cipherKey = getCipherKey(ENCRYPTED_PLUGIN_SALT + NETLIFY_ENCRYPT_KEY);
    const readStream = fs.createReadStream(sourceFilePath, { start: 16 });
    const decipher = crypto.createDecipheriv('aes256', cipherKey, initVect);
    const unzip = zlib.createUnzip();
    const destination = testdecrypt
      ? path.join('.testdecrypt', destinationfilePath)
      : destinationfilePath;
    ensureDirectoryExistence(destination);
    const writeStream = fs.createWriteStream(destination);
    readStream
      .pipe(decipher)
      .pipe(unzip)
      .pipe(writeStream);
  });
}

// util utils
function getCipherKey(password) {
  return crypto
    .createHash('sha256')
    .update(password)
    .digest();
}
// https://stackoverflow.com/questions/13542667/create-directory-when-writing-to-file-in-node-js
function ensureDirectoryExistence(filePath) {
  var dirname = path.dirname(filePath);
  if (fs.existsSync(dirname)) {
    return true;
  }
  ensureDirectoryExistence(dirname);
  fs.mkdirSync(dirname);
}


encrypt.js
#!/usr/bin/env node
'use strict';

const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const zlib = require('zlib');
const chalk = require('chalk');
const { Transform } = require('stream');

const rimraf = require('rimraf');
const sade = require('sade');

const ENCRYPTED_PLUGIN_SALT = 'ENCRYPTED_PLUGIN_SALT';
const { NETLIFY_ENCRYPT_KEY } = process.env;

// for crypto stuff later
class AppendInitVect extends Transform {
  constructor(initVect, opts) {
    super(opts);
    this.initVect = initVect;
    this.appended = false;
  }

  _transform(chunk, encoding, cb) {
    if (!this.appended) {
      this.push(this.initVect);
      this.appended = true;
    }
    this.push(chunk);
    cb();
  }
}

sade('encrypt <regex>', true)
  .version('0.0.1')
  .describe('Encrypt files. Expects a linked site.')
  .example('encrypt secretcontent/**/*.md')
  .action((regex, opts) => {
    if (typeof NETLIFY_ENCRYPT_KEY === 'undefined') {
      console.error(
        'must define NETLIFY_ENCRYPT_KEY to use netlify-plugin-encrypted-files'
      );
      process.exit(1);
    }
    // Program handler
    const filePaths = [regex, ...opts._];
    if (!filePaths.length) throw new Error('no filePaths specified!');

    // prepare encrypt folder
    if (fs.existsSync('.encrypted'))
      console.log('clearing ' + '.encrypted') || rimraf.sync('.encrypted');
    fs.mkdirSync('.encrypted');

    filePaths.forEach((filePath) => {
      console.log(`encrypting ${chalk.blue(filePath)}`);
      encrypt(filePath);
    });
  })
  .parse(process.argv);

// core crypto stuff
// https://medium.com/@brandonstilson/lets-encrypt-files-with-node-85037bea8c0e
function encrypt(file) {
  // Generate a secure, pseudo random initialization vector.
  const initVect = crypto.randomBytes(16);

  // Generate a cipher key from the password.
  const CIPHER_KEY = getCipherKey(ENCRYPTED_PLUGIN_SALT + NETLIFY_ENCRYPT_KEY);
  const readStream = fs.createReadStream(file);
  const gzip = zlib.createGzip();
  const cipher = crypto.createCipheriv('aes256', CIPHER_KEY, initVect);
  const appendInitVect = new AppendInitVect(initVect);
  // Create a write stream with a different file extension.

  let encryptedPath = Buffer.from(file).toString('base64');
  const destination = path.join('.encrypted', encryptedPath);
  const writeStream = fs.createWriteStream(destination);

  readStream
    .pipe(gzip)
    .pipe(cipher)
    .pipe(appendInitVect)
    .pipe(writeStream);
}

// util utils
function getCipherKey(password) {
  return crypto
    .createHash('sha256')
    .update(password)
    .digest();
}
// // https://stackoverflow.com/questions/13542667/create-directory-when-writing-to-file-in-node-js
// function ensureDirectoryExistence(filePath) {
//   var dirname = path.dirname(filePath)
//   if (fs.existsSync(dirname)) {
//     console.log('clearing ' + dirname)
//     rimraf.sync(dirname)
//     // return true
//   }
//   ensureDirectoryExistence(dirname)
//   fs.mkdirSync(dirname)
// }
